File: cpp/124.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  pair<int, int> pathSum(TreeNode *root){
    if(root == nullptr)
      return make_pair(INT_MIN, 0);
    auto l = pathSum(root->left), r = pathSum(root->right);
    int ans = max(root->val + max(0, l.second) + max(0, r.second), max(l.first, r.first));
    return make_pair(ans, max(root->val, max(root->val + l.second, root->val + r.second)));
  }
  int maxPathSum(TreeNode *root) {
    if(root == nullptr) return 0;
    return pathSum(root).first;
  }
};

int main(){
  auto sol = Solution();
  TreeNode a(1), b(2), c(3);
  a.left = &b; a.right = &c;
  cout << sol.maxPathSum(&a);
}

File: python/124.py
class Solution(object):
    def _max(self, root):
        if not root:
            return 0, 0
        if root.left == root.right == None:
            return root.val, root.val

        l1, l2 = self._max(root.left)
        l1 = max(l1, 0)
        r1, r2 = self._max(root.right)
        r1 = max(r1, 0)
        if root.left is None:
            return root.val + r1, max(r2, root.val + r1)
        elif root.right is None:
            return root.val + l1, max(l2, root.val + l1)
        else:
            return root.val + max(l1, r1), max(max(l2, r2), l1 + root.val + r1)

    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self._max(root)[1]
